### PR TITLE
heif: add CreateCopy support (WIP)

### PR DESCRIPTION
**Pull request by bradh**
## What does this PR do?

Implements CreateCopy for the heif driver. The implementation uses libheif.

This implementation also includes some cleanups to existing code. That could be split out into another PR if desired.

## What are related issues/pull requests?

N/A.

## Tasklist

 - [x] Add virtual writer
 - [ ] Add support for != 8 bits
 - [ ] Add support for YUV
 - [ ] Add creation options for libheif encoding options
 - [ ] Add test case for AV1 encoding
 - [x] Add test case for JPEG encoding
 - [x] Add test case for JPEG2000 encoding
 - [ ] Add test case for uncompressed encoding.
 - [ ] remove hack for driver check (https://github.com/OSGeo/gdal/pull/8907/commits/3ac750b1175356e8077b72a207db11f603010186#diff-4462bd609cd228478ce2ea09d926c0edecf4b06c3033f71540cf0fa86cbd664aR311)
 - [ ] Decide on default quality level (50 might be a bit low)
 - [ ] Add documentation - multiple creation options need to be added
 - [ ] Updated Python API documentation (swig/include/python/docs/)
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

## Environment

N/A.

### CODE REVIEW

File: frmts/heif/heifdataset.cpp [hunk 1]
```diff
@@ -700,5 +654,9 @@ void GDALRegister_HEIF()
 
     poDriver->pfnOpen = GDALHEIFDataset::Open;
 
+#ifdef HAS_CUSTOM_FILE_WRITER
```
Comment (rouault) [inline frmts/heif/heifdataset.cpp:657]: HEIFDriverSetCommonMetadata() should set GDAL_DCAP_CREATECOPY when HAS_CUSTOM_FILE_WRITER i set,  for deferred loaded plugins (https://gdal.org/development/rfc/rfc96_deferred_plugin_loading.html)

File: frmts/heif/heifdrivercore.cpp [hunk 2]
```diff
@@ -42,33 +42,33 @@ int HEIFDriverIdentifySimplified(GDALOpenInfo *poOpenInfo)
     if (poOpenInfo->nHeaderBytes < 12 || poOpenInfo->fpL == nullptr)
         return false;
 
+#if LIBHEIF_HAVE_VERSION(1, 4, 0)
+    const auto res =
```
[the remark this answers is not in the document]
Comment (rouault) [reply frmts/heif/heifdrivercore.cpp:46]: > Based on an initial review of the RFC, I understand this is a filter that should return true for all supported files. That is, it is probably better to return a false positive (which will delay startup) than a false negative (which will cause the file not to be loaded), within reason.

yes.
If there are not-sure situations, and mostly if the file might be opened by another driver, you can also return GDAL_IDENTIFY_UNKNOWN, and the Open() method with be invoked.

File: frmts/heif/heifdatasetcreatecopy.cpp [hunk 3]
```diff
@@ -0,0 +1,227 @@
+/******************************************************************************
+ *
+ * Project:  HEIF Driver
+ * Author:   Even Rouault <even.rouault at spatialys.com>
+ *
+ ******************************************************************************
+ * Copyright (c) 2023, Even Rouault <even.rouault at spatialys.com>
```
[the remark this answers is not in the document]
Comment (bradh) [reply frmts/heif/heifdatasetcreatecopy.cpp]: Fixed.

File: frmts/heif/heifdataset.cpp (hunk 1 above)
Comment (bradh) [reply frmts/heif/heifdataset.cpp:657]: Implemented, to be tested.

Comment (bradh) [reply frmts/heif/heifdataset.cpp:657]: Tested

**Comment by bradh** [comment]
If there are feedback items on what is currently implemented, that would be appreciated.

File: autotest/gdrivers/heif.py [hunk 4]
```diff
@@ -158,3 +161,136 @@ def test_heif_subdatasets():
...
+heif_codecs = ["HEIF", "JPEG", "JPEG2000"]
+
+
+@pytest.mark.parametrize("codec", heif_codecs)
+def test_heif_create_copy(codec):
+    tempfile = "tmp/test_heif_create_copy_" + codec + ".hif"
+    try:
+        os.remove(tempfile)
+    except OSError:
+        pass
```
Comment (rouault) [inline autotest/gdrivers/heif.py:245]:
```suggestion
def test_heif_create_copy(tmp_path, codec):
    tempfile = str(tmp_path / "test_heif_create_copy_" + codec + ".hif")
```

and same in below similar occurences

File: frmts/heif/heifdatasetcreatecopy.cpp [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+    heif_writer writer;
+    writer.writer_api_version = 1;
+    writer.write = VFS_WriterCallback;
+    heif_context_write(ctx, &writer, fp);
+    VSIFCloseL(fp);
+
+    heif_context_free(ctx);
+
+    GDALPamDataset *poDS = (GDALPamDataset *)GDALOpen(pszFilename, GA_ReadOnly);
+    return poDS;
```
Comment (rouault) [inline frmts/heif/heifdatasetcreatecopy.cpp:264]:
```suggestion
    return GDALDataset::Open(pszFilename);
```

**Review by rouault — COMMENTED** [review]
a few comments, but looks good. Would need to be rebased on top of latest master to get fresh CI. Do you feel like this PR could be merged as it (with potential future PRs), or is there some extra work required?

**Comment by bradh** [comment]
> a few comments, but looks good. Would need to be rebased on top of latest master to get fresh CI. 

Thanks.

> Do you feel like this PR could be merged as it (with potential future PRs), or is there some extra work required?

I would not want to merge it without the test cases and documentation, or with the driver check hack. The other parts could probably wait for future PRs.

File: doc/source/drivers/raster/heif.rst [hunk 6]
```diff
@@ -31,6 +34,26 @@ Driver capabilities
 
 .. supports_virtualio:: if libheif >= 1.4
 
+Creation support
```
Comment (rouault) [inline doc/source/drivers/raster/heif.rst:37]: to fix "AssertionError: Driver HEIF declares DCAP_CREATECOPY but doc does not!" in https://github.com/OSGeo/gdal/actions/runs/9139007180/job/25130835649?pr=8907:

```suggestion

.. supports_createcopy::

Creation support
```

Comment (bradh) [reply doc/source/drivers/raster/heif.rst:37]: Thanks. Already fixed in a separate work branch. Still not sure what is going on with the bus error on the Alpine builds....

Comment (rouault) [reply doc/source/drivers/raster/heif.rst:37]: > on the Alpine builds....

perhaps a issue specific with the libheif version they ship

A way to reproduce locally is to use the Docker alpine:edge image

docker run --name gdal_alpine -it -v /path/to/your/GDAL/checkout:/gdal alpine:edge

execute the apk add line at https://github.com/OSGeo/gdal/blob/master/.github/workflows/alpine/Dockerfile.ci#L3
cd /gdal
python3 -m pip install --break-system-packages -U -r autotest/requirements.txt
mkdir build_alpine
cd build_alpine
do the build as in https://github.com/OSGeo/gdal/blob/master/.github/workflows/alpine/build.sh#L16 (drop the line "-DADD_EXTERNAL_DEFERRED_PLUGIN_FOO=/tmp/foo.cpp" to simplify things)

**Comment by bradh** [comment]
I'll rework this and submit as a new PR.

**Pull request closed without merging**